system/jlib/jmetrics.hpp

```diff
@@ -0,0 +1,202 @@
+    limitations under the License.
+############################################################################## */
+
+#pragma once
+
+#include <string>
+#include <utility>
+#include <vector>
+#include <map>
+#include <atomic>
+#include <chrono>
+#include <mutex>
+#include <memory>
+#include <unordered_set>
+#include "jiface.hpp"
+#include "jptree.hpp"
+
+
+namespace hpccMetrics {
+
+class MetricsReporter;
+
+MetricsReporter jlib_decl &queryMetricsReporter();
+
+/*
+ * Enumerates the metric type.
+ */
+enum MetricType
+{
+    METRICS_COUNTER,
+    METRICS_GAUGE
+};
+
+
+/*
+ * IMetric
+ *
+ * Interface defining a metric
+ */
+interface IMetric
+{
+    /*
+     * Returns the metric name
+     */
+    virtual const std::string &queryName() const = 0;
+
+    /*
+     * Returns metric description
+     */
+    virtual const std::string &queryDescription() const = 0;
+
+    /*
+     * Returns the metric type.
+     */
+    virtual MetricType queryMetricType() const = 0;
+
+    /*
+     * Get current measurement
+     */
+    virtual uint64_t queryValue() = 0;
+};
+
+
+/*
+ * Concrete base class implementation of the IMetric interface. All metrics inherit
+ * from this class.
+*/
+class jlib_decl Metric : public IMetric
+{
+public:
+    virtual ~Metric() = default;
+    const std::string &queryName() const override { return name; }
+    const std::string &queryDescription() const override { return description; }
+    MetricType queryMetricType() const override { return metricType; }
+    uint64_t queryValue() override { return value; }
+
+protected:
+    // No one should be able to create one of these
+    Metric(const char *_name, const char *_desc, MetricType _metricType) :
+            name{_name},
+            description{_desc},
+            metricType{_metricType} { }
+
+protected:
+    std::string name;
+    std::string description;
+    MetricType metricType;
+    std::atomic<uint64_t> value{0};
+};
+
+/*
+ * Metric used to count events. Count is a monotonically increasing value
+ */
+class jlib_decl CounterMetric : public Metric
+{
+public:
+    CounterMetric(const char *name, const char *description) :
+            Metric{name, description, MetricType::METRICS_COUNTER}  { }
+    void inc(uint64_t val)
+    {
+        value.fetch_add(val);
+    }
+};
+
+
+/*
+ * Metric used to track the current state of some internal measurement.
+ */
+class jlib_decl GaugeMetric : public Metric
+{
+public:
+    GaugeMetric(const char *name, const char *description) :
+        Metric{name, description, MetricType::METRICS_GAUGE}  { }
+
+    /*
+     * Update the value as indicated
+     */
+    void add(int64_t delta)
+    {
+        value += delta;
+    }
+
+    /*
+     * Set the value
+     */
+    void set(int64_t val)
+    {
+        value = val;
+    }
+};
+
+
+
+class jlib_decl MetricSink
+{
+public:
+    virtual ~MetricSink() = default;
+    virtual void startCollection(MetricsReporter *pReporter) = 0;
```

Comment:
minor: This looks like it should be passed to the getSinkInstance() function, but I wouldn't change for this PR.